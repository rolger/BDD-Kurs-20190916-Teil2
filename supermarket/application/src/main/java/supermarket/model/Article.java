package supermarket.model;

public class Article {
    private final ProductCode id;
    private final double quantity;

    public Article(ProductCode id, double quantity) {
        this.id = id;
        this.quantity = quantity;
    }

    public ProductCode getId() {
        return id;
    }

    public double getQuantity() {
        return quantity;
    }

}
