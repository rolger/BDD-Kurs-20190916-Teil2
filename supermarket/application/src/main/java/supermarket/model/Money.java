package supermarket.model;

import java.math.BigDecimal;
import java.util.Objects;

public class Money {

    private final BigDecimal value;

    public Money(int value) {
        this(new BigDecimal(value));
    }

    public Money(BigDecimal value) {
        this.value = value;
    }

    public Money(double value)  {
        this(new BigDecimal(value));
    }

    public Money add(Money other) {
        return new Money(value.add(other.value));
    }

    public Money multiply(double multiplicator) {
        return new Money(value.multiply(new BigDecimal(multiplicator)));
    }

    public Money percentage(int p) {
        return new Money(value.multiply(new BigDecimal(p)).divide(new BigDecimal(100)));
    }

    public BigDecimal asBigDecimal() {
        return value;
    }

    public String format() {
        return String.format("%.2f", value);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Money money = (Money) o;
        return Objects.equals(value, money.value);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value);
    }
}
