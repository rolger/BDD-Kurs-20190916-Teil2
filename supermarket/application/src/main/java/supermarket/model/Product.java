package supermarket.model;

public class Product {
    private final ProductCode code;
    private final String name;
    private final Money basePrice;
    private final Unit unit;

    public Product(ProductCode number, String name, Money basePrice) {
        this(number, name, Unit.PIECE, basePrice);
    }

    public Product(ProductCode number, String name, Unit unit, Money basePrice) {
        this.code = number;
        this.name = name;
        this.unit = unit;
        this.basePrice = basePrice;
    }

    public ProductCode getCode() {
        return code;
    }

    public String getName() {
        return name;
    }

    public Money getBasePrice() {
        return basePrice;
    }

    public Unit getUnit() {
        return unit;
    }
}
