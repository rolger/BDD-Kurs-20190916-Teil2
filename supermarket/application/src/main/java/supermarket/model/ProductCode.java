package supermarket.model;

import java.util.Objects;

public class ProductCode {

    private String code;

    public static ProductCode buildFromBarcode(String code) {
        return new ProductCode(code);
    }

    public ProductCode(String code) {
        if (isNullOrWhiteSpace(code))
            throw new IllegalArgumentException("The code must not be empty.");
        if (!containsJustDigits(code))
            throw new IllegalArgumentException("The code must consist only if digits but was "+ code);
        if (code.length() != 5)
            throw new IllegalArgumentException("The code must be 5 characters long but was " + code);

        this.code = code;
    }

    private boolean isNullOrWhiteSpace(String text) {
        return text == null || text.isEmpty();
    }
    private static boolean containsJustDigits(String text) {
        return text.matches("[0-9]*");
    }

    public String getCode() {
        return code;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ProductCode that = (ProductCode) o;
        return Objects.equals(code, that.code);
    }

    @Override
    public int hashCode() {
        return Objects.hash(code);
    }
    
}