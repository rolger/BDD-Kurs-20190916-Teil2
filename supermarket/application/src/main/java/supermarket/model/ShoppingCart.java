package supermarket.model;

import java.util.ArrayList;
import java.util.List;

public class ShoppingCart {

	private final List<Article> articles = new ArrayList<>();

	public void addArticle(ProductCode code, double quantity) {
		articles.add(new Article(code, quantity));
	}

	public List<Article> getArticles() {
		return new ArrayList<>(articles);
	}

}
