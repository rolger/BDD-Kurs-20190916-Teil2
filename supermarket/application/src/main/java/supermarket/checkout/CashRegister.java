package supermarket.checkout;

import supermarket.model.*;

import java.util.List;

public class CashRegister {

    private final ICatalog catalog;

    public CashRegister(ICatalog catalog) {
        this.catalog = catalog;
    }

    Receipt checkoutArticlesFrom(ShoppingCart theCart) {
        Receipt receipt = new Receipt();

        List<Article> articles = theCart.getArticles();
        for (Article article : articles) {
            final Product product = catalog.getProductByNumber(article.getId());
            Money price = product.getBasePrice().multiply(article.getQuantity());
            receipt.addItem(product, article.getQuantity(), price);
        }

        return receipt;
    }

}
