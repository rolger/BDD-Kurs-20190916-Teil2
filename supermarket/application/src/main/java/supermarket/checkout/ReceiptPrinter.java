package supermarket.checkout;

import org.apache.commons.lang3.StringUtils;
import supermarket.model.Receipt;
import supermarket.model.ReceiptItem;

import java.time.LocalDate;

import static java.lang.String.format;
import static org.apache.commons.lang3.StringUtils.repeat;

public class ReceiptPrinter {

    public static final int COLUMNS = 40;
    public static final String TOTAL_TITLE = "Total :";

    private final LocalDate receiptDate;
    private final String shopName;

    public ReceiptPrinter(String shopName, LocalDate receiptDate) {
        this.shopName = shopName;
        this.receiptDate = receiptDate;
    }

    public String printReceipt(Receipt receipt) {
        StringBuilder receiptBuilder = new StringBuilder();

        printHeader(receiptBuilder);
        printLines(receipt, receiptBuilder);
        printFooter(receipt, receiptBuilder);

        return receiptBuilder.toString();
    }

    private void printHeader(StringBuilder receiptBuilder) {
        receiptBuilder.append(withCenterPadding(shopName));
        nextLine(receiptBuilder);
        nextLine(receiptBuilder);
        receiptBuilder.append("Your purchase from " + receiptDate);
        nextLine(receiptBuilder);
        separatorLine(receiptBuilder);
    }

    private void printLines(Receipt receipt, StringBuilder receiptBuilder) {
        receiptBuilder.append(withLeftPadding(0, "EUR"));
        nextLine(receiptBuilder);

        for (ReceiptItem item : receipt.getItems()) {
            String name = item.getText();
            String price = item.getTotalPrice().format();

            receiptBuilder.append(name);
            receiptBuilder.append(withLeftPadding(name.length(), price));
            nextLine(receiptBuilder);
        }
    }

    private void printFooter(Receipt receipt, StringBuilder receiptBuilder) {
        separatorLine(receiptBuilder);
        receiptBuilder.append(TOTAL_TITLE);
        receiptBuilder.append(withLeftPadding(TOTAL_TITLE.length(), receipt.getTotalPrice().format()));
        nextLine(receiptBuilder);
    }

    private void separatorLine(StringBuilder receiptBuilder) {
        receiptBuilder.append(repeat("-", COLUMNS));
        nextLine(receiptBuilder);
    }

    private void nextLine(StringBuilder receiptBuilder) {
        receiptBuilder.append("\n");
    }

    private String withCenterPadding(String string) {
        return repeat(" ", (COLUMNS - string.length()) / 2) + string;
    }

    private String withLeftPadding(int length, String string) {
        return format("%1$" + (COLUMNS - length) + "s", string);
    }

}
