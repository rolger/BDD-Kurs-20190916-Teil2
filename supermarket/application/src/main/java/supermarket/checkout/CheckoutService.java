package supermarket.checkout;

import supermarket.model.ICatalog;
import supermarket.model.Product;
import supermarket.model.ProductCode;
import supermarket.model.Receipt;
import supermarket.model.ShoppingCart;

import java.time.LocalDate;

import static java.lang.String.format;

public class CheckoutService {
    private final String shopName;
    private final ICatalog catalog;
    private final ShoppingCart shoppingCart;

    public CheckoutService(String shopName, ICatalog catalog, ShoppingCart theCart) {
        this.shopName = shopName;
        this.catalog = catalog;
        this.shoppingCart = theCart;
    }

    public void onBarCode(String barcode, double quantity) {
        ProductCode code = new ProductCode(barcode);

        checkIfProductExists(barcode, code);
        shoppingCart.addArticle(code, quantity);
    }

    private void checkIfProductExists(String barcode, ProductCode code) {
        final Product product = catalog.getProductByNumber(code);
        if (product == null) {
            throw new IllegalArgumentException(format("The product with code %s doesn't exist in the catalog.", barcode));
        }
    }

    public String printReceipt(final CashRegister cashRegister, LocalDate receiptDate) {
        Receipt receipt = cashRegister.checkoutArticlesFrom(shoppingCart);
        return new ReceiptPrinter(shopName, receiptDate).printReceipt(receipt);
    }

}
