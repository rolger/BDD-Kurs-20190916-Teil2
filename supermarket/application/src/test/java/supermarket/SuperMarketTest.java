package supermarket;

import org.junit.Before;
import org.junit.Test;
import supermarket.checkout.CashRegister;
import supermarket.checkout.CheckoutService;
import supermarket.model.Money;
import supermarket.model.Product;
import supermarket.model.ProductCode;
import supermarket.model.ShoppingCart;

import java.time.LocalDate;

import static org.approvaltests.Approvals.verify;
import static supermarket.model.ProductCode.buildFromBarcode;

public class SuperMarketTest {
    public static final String SHOP_NAME = "BEST SHOPPING";

    private static final ProductCode apples = buildFromBarcode("00001");
    private static final ProductCode bread = buildFromBarcode("00002");
    private static final ProductCode milk = buildFromBarcode("00003");

    private static final LocalDate TRAINING_DATE = LocalDate.of(2019, 9, 17);

    private MemoryProductCatalog catalog;
    private CheckoutService checkoutService;

    @Before
    public void setUp() {
        catalog = new MemoryProductCatalog();
        catalog.add(new Product(apples, "Apfel", new Money(0.9)));
        catalog.add(new Product(bread, "Schwarzbrot", new Money(2.49)));
        catalog.add(new Product(milk, "Milch", new Money(1.01)));

        checkoutService = new CheckoutService(SHOP_NAME, catalog, new ShoppingCart());
    }

    @Test
    public void buyOneArticle() {
        checkoutService.onBarCode(bread.getCode(), 2);

        verify(checkoutService.printReceipt(new CashRegister(catalog), TRAINING_DATE));
    }

    @Test
    public void buyThreeArticlesWithoutDiscount() {
        checkoutService.onBarCode(apples.getCode(), 0.73);
        checkoutService.onBarCode(bread.getCode(), 1);
        checkoutService.onBarCode(milk.getCode(), 2);

        verify(checkoutService.printReceipt(new CashRegister(catalog), TRAINING_DATE));
    }
}
