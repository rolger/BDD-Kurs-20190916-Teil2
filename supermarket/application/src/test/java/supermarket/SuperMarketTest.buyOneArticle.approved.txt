             BEST SHOPPING

Your purchase from 2019-09-17
----------------------------------------
                                     EUR
Schwarzbrot                         4,98
----------------------------------------
Total :                             4,98
